Introduce AnnotatedIOBuffer

tecosaur opened the pull request:
This allows for styled content to be constructed incrementally, without resorting to repeated concatenation. It operates very similarly to IOContext, just with a special `write` method and specifically wrapping an IOBuffer.

-----

While I consider this functionality generally useful, I also have a specific use in `Markdown` in mind for this. For no it's not exported or public API, but I think we'd probably want to talk about making it public API in the future.

For `Base` itself though, I think it's important to note that this lets us pass an `AnnotatedIOBuffer` to `show` calls, and thus truncate printed `AnnotatedStrings` without the issues of accidentally removing terminating escape codes etc. that we currently have (see #45521 for an example).

With the way `printstyled` works, this benefit won't be seen immediately, but I see it as part of building the path to a nicer, less headache-inducing future.

tecosaur committed:
Introduce AnnotatedIOBuffer

This allows for styled content to be constructed incrementally, without
resorting to repeated concatenation. It operates very similarly to
IOContext, just with a special `write` method and specifically wrapping
an IOBuffer.

LilithHafner commented:
This CI failure:
```
ambiguous                                        (3) \|         failed at 2023-10-21T12:07:05.619
  | Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-b67d93b47d/share/julia/test/ambiguous.jl:174
  | Expression: sig1 === sig2
  | Evaluated: write(io::Base.AnnotatedIOBuffer, x::AbstractString) @ Base strings/annotated.jl:433 === write(io::IO, s::Union{SubString{String}, String}) @ Base strings/io.jl:248
  |  
  | Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-b67d93b47d/share/julia/test/ambiguous.jl:177
  | Expression: good
```
Means that `write(io::Base.AnnotatedIOBuffer, x::AbstractString)` and `write(io::IO, s::Union{SubString{String}, String})` are ambiguous.

Specifically, calls to `write(io::Base.AnnotatedIOBuffer, s::Union{SubString{String}, String})` will now error.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+    dict::ImmutableDict{Symbol, Any}
```
I suspect that this `dict` field belongs in `IOContext`, not the buffer itself. (e.g. `IOContext{AnnotatedIOBuffer}`).

```julia
julia> dump(IOContext(IOBuffer(), :a => 2))
IOContext{IOBuffer}
  io: IOBuffer
    data: Array{UInt8}((32,)) UInt8[0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  …  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
    reinit: Bool false
    readable: Bool true
    writable: Bool true
    seekable: Bool true
    append: Bool false
    size: Int64 0
    maxsize: Int64 9223372036854775807
    ptr: Int64 1
    mark: Int64 -1
  dict: Base.ImmutableDict{Symbol, Any}
    parent: Base.ImmutableDict{Symbol, Any}
      parent: #undef
      key: #undef
      value: #undef
    key: Symbol a
    value: Int64 2
```

tecosaur replied:
Ah yes, you can just wrap an `AnnotatedIOBuffer` in an IO context. At the time I added this, I was thinking that since this is useful for printing things, it could be nice if an `AnnotatedIOBuffer` could also contain properties like `displaysize`.

Upon reflection, I'm not sure whether this is actually a good idea, but it does seem like something that could be added later if it turns out that we'd like it: So I'm thinking remove the dict and we can cross that bridge later if we want.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    print(io, '(', annio.io.size, " bytes)")
+end
+
+position(io::AnnotatedIOBuffer) = position(io.io)
+lock(io::AnnotatedIOBuffer) = lock(io.io)
+unlock(io::AnnotatedIOBuffer) = unlock(io.io)
+get(io::AnnotatedIOBuffer, key, default) = key == :color || get(io.dict, key, default)
+displaysize(io::AnnotatedIOBuffer) =
+    if haskey(io.dict, :displaysize) io.dict[:displaysize]::Tuple{Int,Int} else displaysize(io.io) end
```
Are you following some sort of documented interface for how to define a fully featured IO type, or is this just a classic julian undocumented set of methods that happen to do the trick? (It's not your responsibility to document how to define a new IO type, but if it is documented somewhere, I'd love to know about it)

tecosaur replied:
I looked at the implementation of `IOContext` (since this works similarly), and then this is a case of good ol' copy + paste + tweak (I'm not aware of any documented interface, but maybe I missed something?).

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    annot = copy(aiob.annotations)
+    empty!(aiob.annotations)
+    seekstart(aiob.io)
+    str, annot
+end
+
+AnnotatedString((str, annots)::Tuple{<:AbstractString, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}}) =
+    AnnotatedString(str, annots)
```
This seems like a strange API decision; why not return an `AnnotatedString`?

tecosaur replied:
This enables `AnnotatedString(take!(::AnnotatedIOBuffer))`. I considered having `AnnotatedString(::AnnotatedIOBuffer)`, but I don't like that it mutates the state of the `AnnotatedIOBuffer` within the constructor, that feels icky to me.

tecosaur replied:
It occurs to me that maybe `read(::AnnotatedIOBuffer, AnnotatedString)` may also be reasonable.

LilithHafner replied:
One worry I have about the current approach is that it exposes the annotations field which is otherwise internal.

I like `read(::AnnotatedIOBuffer, AnnotatedString)`

`take!(::IOBuffer)` returns a vector, not an AbstractString. It seems reasonable to be consistent with that behavior, but also idk how to return a vector analogue that contains annotation information. I don't love `AnnotatedVector`—that feels like a bit much.

tecosaur replied:
Changed :)

LilithHafner reviewed (commented):
This seems like a good idea. However, I think it makes sense to include at least one motivating usage of this type in this PR (or a linked PR that depends on it) because I don't like introducing additional complexity that does not have any impact on publicly accessible functionality.

In addition to avoiding premature addition of complexity (which I doubt this is) the request to add a motivating use case is a nice principle to follow because having a use of `AnnotatedIOBuffer` gives context for the API design decisions of the type.

tecosaur commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    end
+    write(io.io, astr)
+end
+write(io::AnnotatedIOBuffer, achr::AnnotatedChar) = write(aio, AnnotatedString(achr))
```
With PRs that haven't received feedback yet (like this one until very recently), the model I'm currently running with for new-feature PRs is:
1. Propose design
2. Await comments on design
3. Once design seems ok, add tests

Happy to tweak this if it makes things easier though.

LilithHafner replied:
Seems reasonable.

tecosaur commented:
Thanks for the comments Lilith. Regarding adding a usage example, there are three PRs I have in mind currently that need this: improving the `join` method for annotated strings, using `AnnotatedIOBuffer` with `show`/`showerror`, and a PR to the `Markdown` stdlib.

The latter two will be more work I suspect, so I'll see if I can whip up the `join` PR in the near future.

JeffBezanson commented on base/show.jl:
```diff
@@ -429,6 +429,9 @@ function show_circular(io::IOContext, @nospecialize(x))
 end
 
+AnnotatedIOBuffer(io::IOContext{IOBuffer}) = AnnotatedIOBuffer(io.io, io.dict)
+IOContext(io::AnnotatedIOBuffer) = IOContext(io.io, io.dict)
```
I don't see how these are supposed to work?

tecosaur replied:
Those are supposed to be gone in response to https://github.com/JuliaLang/julia/pull/51807#discussion_r1372536999. I'll correct this oversight shortly.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,51 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+annotations of any `AnnotatedString`s written to `io` and otherwise acts like
+`read(io::IO, String)`.
+"""
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{>:String}})
+    str = String(take!(io.io))
+    annots = copy(io.annotations)
+    empty!(io.annotations)
+    seekstart(io.io)
+    AnnotatedString(str, annotss)
+end
```
I'm not super familiar with how to properly use the angry face emoji `>:`. Is the point here that this will be dispatched to on `read(::AnnotatedIOBuffer, AnnotatedString{String})` and `read(::AnnotatedIOBuffer, AnnotatedString{AbstractString})` and `read(::AnnotatedIOBuffer, AnnotatedString{Any})` but not for `read(::AnnotatedIOBuffer, AnnotatedString{MyString})` where `MyString <: AbstractString`?

Would this correctly support all parameterizations of AnnotateString:

```suggestion
function read(io::AnnotatedIOBuffer, ::Type{T}) where T <: AnnotatedString
    str = String(take!(io.io))
    annots = copy(io.annotations)
    empty!(io.annotations)
    seekstart(io.io)
    T(str, annotss)
end
```

tecosaur replied:
Well, we get an `AnnotatedString{String}` from this, so I think it would be reasonable to dispatch on all supertypes (of the `String` type parameter), which is really just:

```
julia> AnnotatedString{String} <: AnnotatedString{>:String}
true

julia> AnnotatedString{AbstractString} <: AnnotatedString{>:String}
true
```

However, this seems to prevent dispatch on the no-type-annotation form, so a `Union` might also be needed? I think for brevity we want `read(io, AnnotatedString)` to work.

LilithHafner replied:
Oops, my initial suggestion was bad.
```suggestion
read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}}) where T
    str = read(io.io, T)
    annots = copy(io.annotations)
    empty!(io.annotations)
    seekstart(io.io)
    AnnotatedString(str, annotss)
end
```
IIUC, this would make an `AnnotatedString{T}` when passed a `AnnotatedString{T}` as input, provided you can read a `T` from the io object.

tecosaur replied:
I'm not sure how worthwhile this would be, given `T <: AbstractString` and that currently the only `T` for which this read method is implemented is `String`.

```

julia> methods(read, Tuple{IO, Type{<:AbstractString}})
# 4 methods for generic function "read" from Base:
 [1] read(stream, ::Type{Union{}}, slurp...; kwargs...)
     @ io.jl:222
 [2] read(io::IO, ::Type{T}) where T<:Enum
     @ Base.Enums Enums.jl:22
 [3] read(s::IO, ::Type{String})
     @ io.jl:1017
 [4] read(s::IO, T::Type)
     @ io.jl:1018
```

tecosaur replied:
The approach has been updated in subsequent conversation with Jamerson.

tecosaur commented:
For a basic example of this being useful, perhaps https://github.com/JuliaLang/StyledStrings.jl/pull/12 is a decent starting point? The `join` PR might not come till the weekend.

LilithHafner commented:
From triage: Jeff is afraid that everyone will have to use annotated things for all IO in Base (and the whole ecosystem?) if we go this route.

This is not something that we can talk about as a single issue. We need to discuss the overarching vision of all IO in Julia. This has far reaching consequences and shouldn't be merged as a small thing when it's really a huge thing.

This is not to say that we don't want to do this, but that if we do decide to do this, it should be a broad sweeping decision. Let's talk about this at next triage 1:30 UTC on November 9.

vchuravy commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,52 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: IO
```
Maybe just call this `AnnotatedBufffer`? It's missing seek for really being an `IOBuffer`, and `read` is `take!`

tecosaur replied:
Could do. We could also forward those functions onto the underlying `IOBuffer`.

KristofferC replied:
This is still to be done?

tecosaur commented:
Triage didn't get to this, try again in a month's time :disappointed:

tecosaur commented:
Triage had a chat about this, and while not unanimous was broadly in support of this. We concluded that it would be best to put some of the major point in the discussion here, and give an opportunity for objections/problems we've overlooked to be mentioned. Should nothing (new) come up, this can be merged in a week or so.

Triage conversation points:
- `AnnotatedIOBuffer` acts as a annotated string builder
- `AnnotatedIOBuffer` acts as a way to get structured styling information out of "legacy" code written for `::IO` and using `print` / `printstyled`. This lets you e.g. safely take a substring of a `show(...)` invocation's result.
- `AnnotatedIOBuffer` isn't really conceptually purely `IO`
- Worst case seems to be `get(::IO, :color, false)`-guarded manually printed escape codes will be lost
- You can't really get the "swap in and collect structured styling information" behaviour _without_ subtyping `IO`
- Since this is an opt-in where you want it type approach, this doesn't affect or require rewriting of existing code, other than funky shouldn't-really-be-done-in-the-first-place type stuff needing rewriting when combined with this.
- As a compromise, it seems worth it

tecosaur commented:
It you have reasons why you think this shouldn't be merged, it would be really helpful if you could also try to think of other ways of gaining the same benefits that this PR is constructed for :pray:.

tecosaur committed:
Make AnnotatedIOBuffer behave more like IOBuffer

A few extra methods help a lot. It also turns out that we don't actually
need to implement lock/unlock, the fallback implementations are enough.

tecosaur committed:
Tests for AnnotatedIOBuffer

tecosaur commented:
I've just pushed two extra commits. The first make `AnnotatedIOBuffer` act more like `IOBuffer`, and the second is a bunch of tests.

tecosaur commented:
Given the outcome of triage last week, I think this should be able to be merged imminently, barring any last-minute concerns around the implementation that need to be resolved prior to merging.

KristofferC commented:
I'm lacking some "higher level guidance" of when to use this and how? The type itself has no documentation, the only thing with a docstring is the `read` method which references the undocumented type. 

And I feel that e.g. this point:

> From triage: Jeff is afraid that everyone will have to use annotated things for all IO in Base (and the whole ecosystem?) if we go this route. This is not something that we can talk about as a single issue. We need to discuss the overarching vision of all IO in Julia. This has far reaching consequences and shouldn't be merged as a small thing when it's really a huge thing.

is yet to be addressed except with

> Triage had a chat about this, and while not unanimous was broadly in support of this. 

which doesn't address the original worry. Was Jeff in this call? @JeffBezanson could you elaborate on what here made your original discomfort go away (I am interested because I have the exact same thoughts when seeing this)?

KristofferC commented on base/show.jl:
```diff
@@ -429,6 +429,11 @@ function show_circular(io::IOContext, @nospecialize(x))
+# An AnnotatedIOBuffer specialisation that needs to be made
+# here as `annotated.jl` is loaded earlier.
+get(io::IOContext{AnnotatedIOBuffer}, key, default) =
+    if key === :color false else get(io.dict, key, default) end
```
```suggestion
    key === :color ? false : get(io.dict, key, default)
```

vtjnash replied:
It seems a little strange that you are dispatching on the element type here. That is type piracy, and can be unreliable. Can we do this more carefully in the constructor of `IOContext` instead (which should directly inherit the `:color` option by querying AnnnotatedIOBuffer on construction)

tecosaur replied:
Well, we want to make sure that no ANSI codes are printed to an `AnnotatedIOBuffer`. That said. since `get(::IOBuffer, :color, false) = false` maybe this isn't needed at all?

vtjnash replied:
What was the consensus here? Seems like this should be removed?

tecosaur replied:
I'm not sure this was ever discussed enough for there to be a consensus, but I think so. I'll get rid of it for now, should it turn out we want to do anything special to prevent styling control codes from being printed into an `AnnotatedIOBuffer` we can cross that bridge when we come to it.

vtjnash replied:
Styling controls are not wrong either, if the user specifically directed them to be printed into the textual part of the buffer (bypassing the annotations). That may be required to be handled anyways as well, because of the semi-frequent pattern of doing temporary buffers: `print(io, repr(thing; context=io))`. We should eventually try to ensure that usually methods that take as argument a `context isa AnnotatedIO` that those do return an `AnnotatedString` however, to avoid most of those issues.

Seelengrab commented:
> I'm lacking some "higher level guidance" of when to use this and how?

The context here is https://github.com/JuliaLang/julia/issues/50817

tecosaur commented:
That gives some context, but not adding a good docstring to the type itself is an oversight on my part. I'll write one now and (force) push it shortly.

LilithHafner commented:
> Was Jeff in this call?

No

tecosaur commented:
> which doesn't address the original worry.

Triage did explicitly discuss the concern articulated in https://github.com/JuliaLang/julia/pull/51807#issuecomment-1781768196

The response to "Jeff is afraid that everyone will have to use annotated things for all IO in Base (and the whole ecosystem?) if we go this route." is that this is not an implication of this PR, articulated in this point from my triage summary message:

> - Since this is an opt-in where you want it type approach, this doesn't affect or require rewriting of existing code, other than funky shouldn't-really-be-done-in-the-first-place type stuff needing rewriting when combined with this.

It would have been nice to have Jeff on the call, but there's there's not much that can be done about that. I suppose we could wait _another_ month or two and re-visit this on triage, but I'd rather not if possible: it's already been a topic of 2.5 meetings (I'm counting the triage meeting I sat in hoping to discuss this but when it didn't make it as 0.5).

KristofferC commented:
> Was Jeff in this call?

> > No

Okay, while I understand the desire to merge things it feels a bit odd to have a triage call where Jeff says something quite strong like

> This has far reaching consequences and shouldn't be merged as a small thing when it's really a huge thing.

and then just having another one where he cannot attend and then deciding it can now actually be merged as a small thing. That's just my reaction to it looking at it from "the outside".

tecosaur commented:
Ok, so does anybody want to insist on bumping this into another triage call, or can we just merge it?

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+end
+
+isopen(io::AnnotatedIOBuffer) = isopen(io.io)
+close(io::AnnotatedIOBuffer) = close(io.io)
```
```suggestion
close(io::AnnotatedIOBuffer) = close(io.io)
closewritable(io::AnnotatedIOBuffer) = closewritable(io.io)
```

JeffBezanson replied:
`closewrite`?

tecosaur replied:
Added.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    write(io.io, String(astr))
+end
+write(io::AnnotatedIOBuffer, c::AnnotatedChar) = write(io, AnnotatedString(c))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, s)
+write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
```
```suggestion
write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
unsafe_write(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_write(io.io, b, len)
```

Performance will typically be punishingly slow, unless you defined `unsafe_write` and `unsafe_read`. You could instead make this a subtype of `AbstractPipe` instead of directly implementing all of `IO`, since then you only need to define `pipe_reader` and `pipe_writer` and methods that you want to override (e.g. this is how `IOContext` is implemented over `IOBuffer`, to add/override `get` but inherit everything else)

tecosaur replied:
I've now added `unsafe_write` and `unsafe_read` passthrough methods. I've stayed away from `AbstractPipe` because of uncertainty around #51808 and it just generally not being clear what exactly the `AbstractPipe` interface is (what do you need to do? which methods are then provided?).

From looking at the `AbstractPipe` code, it seems like you don't get `seek`, `skip`, or `position` (which I'd think are useful enough to be worth having), so we could maybe subtype that and then just provide those methods? I wonder if it would make more sense to do so after the `AbstractPipe` interface is better documented though.

vtjnash replied:
Yeah, we could consider adding `seek`, `skip`, and `position` that affects the `pipe_reader`. Adding to the docstring would also be helpful. Originally it was basically just an internal helper type because it was a huge annoyance to need to define all those methods you just wrote above and then need to write tests for all of them. But clearly it seems to be pretty broadly useful too for defining other new IO wrappers like this.

vtjnash replied:
Looking again in comparison to AbstractPipe inheritence, you also may be missing forwarding opportunities for:
isreadable, iswritable, buffer_writes, flush, readavailable, bytesavailable, reseteof, copyuntil, readuntil_vector!, readbytes!, peek, wait_readnb, wait_close

Many of these will have generic fallbacks, though they may have less functionality, but are probably fine for IOBuffer as there may not be a specialized version for IOBuffer anyways

tecosaur replied:
Ok, at this point I'll just switch to inheriting from `AbstractPipe`.

A subsequent PR from somebody who has more of a sense of the purpose of it adding/clarifying docs explaining `AbstractPipe` a bit more I think would be well worth it though.

vtjnash replied:
I just merged more docs for that if you look now :)

tecosaur replied:
Oh brilliant! Thanks for that Jameson :)

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+        AnnotatedString(read(io.io, String), annots)
+    end
+end
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{AbstractString}}) = read(io, AnnotatedString{String})
```
```suggestion
read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}} where String<:T<:AbstractString) = read(io, AnnotatedString{String})
```

tecosaur replied:
I don't quite get this. The type restriction on `T` (`String<:T<:AbstractString`) only applies when `T` is `String` or `AbstractString`, no? The `String` case already has a dedicated method, so this method then only applies to `AbstractString`, and is equivalent to the current code, just with a less specialised return type.

Am I missing something?

vtjnash replied:
It would also be any union too, though I guess that isn't necessarily an interesting case to handle

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    end
+end
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{AbstractString}}) = read(io, AnnotatedString{String})
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
```
```suggestion
read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
read(io::AnnotatedIOBuffer, ::Type{AnnotatedChar}) = AnnotatedChar(read(io, Char))
```

(not actually certain if this is the right definition to pair with the `write` above)

tecosaur replied:
This isn't quite right, but I can implement an `read(io::AnnotatedIOBuffer, ::Type{AnnotatedChar{Char}})` that works as an analogue to `read(io, Char)` also grabbing annotations fairly easily.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+end
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{AbstractString}}) = read(io, AnnotatedString{String})
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
+read(io::AnnotatedIOBuffer, x) = read(io.io, x)
```
```suggestion
read(io::AnnotatedIOBuffer, T::Type{<:AnnotatedString}) = T(read(io.io, AnnotatedString{String}))
read(io::AnnotatedIOBuffer, x) = read(io.io, x)
```

tecosaur replied:
I've just had a thought regarding making reading more generic, and will make a comment to the PR on that shortly.

vtjnash replied:
This seems to be the added ambiguity. I am not sure there is a way to write this with fully generic `x::Any`, but possibly would work with `x::Type`?
```suggestion
read(io::AnnotatedIOBuffer, x::Type) = read(io.io, x)
```

vtjnash replied:
Actually, I think this attempted method just conflicts with the usual rule that types should customize their `read` implementation using more primitive methods underlying it (e.g. `unsafe_read` or `read(io, UInt8)`)
```suggestion
```

tecosaur replied:
Oh, interesting. I guess I'll try just deleting that line and see how CI goes :crossed_fingers:

tecosaur replied:
Actually, I think I may be able to get rid of the following three `read` methods too. I'll try it and see what CI thinks.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+read(io::AnnotatedIOBuffer, x) = read(io.io, x)
+# Avoid method ambiguity
+read(io::AnnotatedIOBuffer, s::Type{String}) = read(io.io, s)
+read(io::AnnotatedIOBuffer, b::Type{UInt8}) = read(io.io, b)
```
```suggestion
read(io::AnnotatedIOBuffer, b::Type{UInt8}) = read(io.io, b)
unsafe_read(io::AnnotatedIOBuffer, p::Ptr{UInt8}, len::UInt) = unsafe_read(io.io, p, len)
```

tecosaur replied:
Added, just up in the "generic IOBuffer passthroughs" block.

JeffBezanson commented:
My worry was that we are fundamentally changing our IO model so that you always need to think of it as a byte stream plus annotations, instead of just a byte stream. For example, you can no longer assume that writing to an intermediate IOBuffer gives the same output as writing directly to the destination. Of course that was already kind of true due to `IOContext`, so maybe it is not much different.

Aside from that I am always trying to avoid bloat in Base. Could this go in StyledStrings instead for example? Maybe Base does not need all the nicest features --- after all, if Base has all the nice features it will just encourage people to put more code in Base in the future :)

vtjnash commented:
> you can no longer assume that writing to an intermediate IOBuffer gives the same output

That was one of the issues I noticed with my past try at this (I had called it IOColor at the time), since even trivial cases like writing a String to the buffer might have a different result depending on whether the stream was expecting to be UTF-8 or UTF-32, where some of the APIs (like print) expect to change the format of the bytes to match the underlying stream and some do not expect to do so (like write). That is already a likely problem just with formatting into an intermediate IOBuffer, since we often (but don't always) call `String(take!(buffer))` afterwards, and may directly call `write(io, buffer)` instead, even though those could have slightly different meanings.

tecosaur commented:
Jeff:

> Of course that was already kind of true due to IOContext, so maybe it is not much different.

Yea, you can implement this purely using IOContext, just uglier.

> Aside from that I am always trying to avoid bloat in Base. Could this go in StyledStrings instead for example?

Not if we want #51914, which I think we do, or a few other things.

I think it would be helpful if people taking a look at my PRs from this series could take a look at the diagram in #50817 for context. It's seemed like it would be helpful quite often.

TLDR; There are reasons why I've made this a PR to base.

-----

Jameson:

Thanks for the code review. I'll try to get to your comments in the next few days.

tecosaur committed:
Make AnnotatedIOBuffer reading more generic

While `String` is the only concrete type for which `read(::IOBuffer,
::Type{<:AbstractString})` is defined, is is entirely conceivable that
some other custom string type could define a similar `read` method.

Since making reading an `AnnotatedString` from an `AnnotatedIOBuffer`
more generic is as easy as replacing the hardcoded `String` with a type
parameter, we may as well do so.

tecosaur committed:
Also read AnnotatedChars from an AnnotatedIOBuffer

The `read(::AnnotatedIOBuffer, AnnotatedString)` method is intended as
an approximate analogue to `read(::IOBuffer, String)`. In the same
sense, it makes sense to define `read(::AnnotatedIOBuffer,
AnnotatedChar)` as an analogue to `read(::IOBuffer, Char)`.

tecosaur commented:
I've just gone through Jameson's comments, and added the missing `IOBuffer` passthrough methods (`closewrite`, `unsafe_read`, and `unsafe_write`).

Jameson, you also made some suggestions around making the `read(::AnnotatedIOBuffer, ::Type{AnnotatedString{<:AbstractString}})` methods more generic, which has actually prompted the thought that I can just parameterise the `String` in the `read(::AnnotatedIOBuffer, AnnotatedString{String})` method. That just leaves the two fallback methods for when there's no type parameter (`read(::AnnotatedIOBuffer, AnnotatedString)` and an abstract string type parameter (`read(::AnnotatedIOBuffer, AnnotatedString{AbstractString})`), which I think are still fine just calling the `String`-specialised form (see my reply to your review comment).

Following your request for an equivalent `AnnotatedChar`-reading method, I've also plopped that in as a new commit.

-----

My current understanding is that this received a burst more attention in the last triage meeting, but that the conditional approval from the prior triage meeting still applies, and so this can be merged once Jameson's comments are resolved. Do let me know if this isn't the case.

tecosaur committed:
Update the mention of annotated features in NEWS

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+        start, stop = first(region), last(region)
+        push!(io.annotations, (start+offset:stop+offset, annot))
+    end
+    write(io.io, String(astr))
```
Should this also first erase any existing annotations for this region in the `io`?

tecosaur replied:
Probably I think. This is what an implementation of that looks like:

```julia
if !eof(io)
    # If we are overwriting an existing span in the AnnotatedIOBuffer,
    # clear out any overlapping pre-existing annotations.
    span = offset:offset+ncodeunits(astr)
    filter!(((range, _),) -> first(span) <= first(range) && last(range) <= last(span), io.annotations)
    extras = Tuple{UnitRange{Int}, Pair{Symbol, Any}}[]
    for i in eachindex(io.annotations)
        range, annot = io.annotations[i]
        # Test for partial overlap
        if first(span) <= first(range) <= last(span) || first(span) <= last(range) <= last(span)
            io.annotations[i] = (if first(range) < first(span)
                                     first(range):first(span)-1
                                 else last(span)+1:last(range) end, annot)
            # If `span` fits exactly within `range`, then we've only copied over
            # the beginning overhang, but also need to conserve the end overhang.
            if first(range) < first(span) && last(span) < last(range)
                io.annotations[i] = (first(range):first(span)-1, annot)
                push!(extras, (last(span)+1:last(range), annot))
            end
        end
        # Insert any extra entries in the appropriate position
        for entry in extras
            indices = searchsorted(io.annotations, (first(entry),), by=first)
            splice!(s.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
        end
    end
end
```

tecosaur replied:
Implemented with tests.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+otherwise acts like `read(io::IO, String)`.
+"""
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}}) where {T <: AbstractString}
+    if eof(io)
+        seekstart(io.io)
```
This is invalid for IO

tecosaur replied:
It's a pity this, because what's really wanted is an `AnnotatedString` analogue for `String(take!(io))`. I suppose `read(seekstart(io), AnnotatedString)` will have to do.

tecosaur replied:
No longer done.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+"""
+    read(io::AnnotatedIOBuffer, AnnotatedString)
+
+Read `io` as an `AnnotatedString`, jumping to the start if `eof(io)` holds.
```
```suggestion
Read `io` as an `AnnotatedString`.
```

It is not valid for a method to define a function to mean something completely different from the method it overrides.

tecosaur replied:
The method no longer does this.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
+
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedChar{T}}) where {T <: AbstractChar}
+    char = read(io.io, T)
+    pos = position(io)
```
```suggestion
    pos = position(io)
    char = read(io.io, T)
```
?

tecosaur replied:
Since the annotation range should cover the whole character, and the buffer is zero-indexed unlike the annotations, getting the position this way means we get the 1-based index of the last codepoint of the character. It's probably clearer if I reverse this though, so I'll do that and filter on the first codepoint instead.

tecosaur replied:
Done.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+unsafe_write(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_write(io.io, b, len)
+unsafe_read(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_read(io.io, b, len)
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
```
You probably also want to add a similar method to copy the remainder from one annotated IO to another:
```julia
function write(to::AnnotatedIOBuffer, from::AnnotatedIOBuffer)
    fromoffset = position(from)
    tooffset = position(to)
    nb = write(to.io, from.io)
    append!(to.annotations, (
        (clamp(region .- (tooffset - fromoffset)), annot)
        for (region, annot) in from.annotations
        if last(region) > fromoffset)) # approximation only, since I am not sure about the comparisons or clamping of the start of the region
    return nb
end
```

tecosaur replied:
Why not, could be nice to have. I'll push an initial implementation, and then add some tests in a bit.

tecosaur commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+        AnnotatedString(read(io.io, T), copy(io.annotations))
+    else
+        start = position(io.io)
+        annots = map(((range, val),) -> (max(1, first(range) - start):last(range)-start, val),
+                     filter(((range, _),) -> last(range) > start,
+                            io.annotations))
```
Ah yep, that's nice. Thanks!

tecosaur committed:
Writing from one AnnotatedIOBuffer to another

tecosaur reviewed (commented):
I think that's everything addressed?

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    srcpos = position(src)
+    offset = destpos - srcpos
+    nb = write(dest.io, src.io)
+    # Clear out annotations that only start after `position(dest)`,
+    filter!(((region, _),) -> first(region) <= destpos, dest.annotations)
```
Should this put back afterwards any annotations that end after the `position(destpos)`, so that it only overwrites the mid-portion of the annotations? (In lisp, I would say to use `separate!` instead of `filter!`, but I don't know if that exists in Julia)

vtjnash replied:
```suggestion
    # Preserve any annotations that end after `position(dest)`
    destpos2 = position(dest)
    endannotations = filter(((region, _),) -> last(region) >= destpos2, dest.annotations)
    # Truncate those existing annotation to start after `destpos2`
    map!(((region, annot),) -> (max(first(region), destpos2):last(region), annot),
         endannotations, endannotations)
    # Clear out annotations that only start after `position(dest)`,
    filter!(((region, _),) -> first(region) <= destpos, dest.annotations)
```

tecosaur replied:
Ah yes, that would seem sensible. That said, there's a little more too it since annotations that go after the end of the write may need to have their region trimmed at the start too. This logic is already in the `write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})` method, so I've extracted the logic into a helper function `_clear_annotations_in_region!` that can be used here too.

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+    append!(dest.annotations,
+            ((max(1, first(region) + offset):last(region)+offset, annot)
+             for (region, annot) in src.annotations if first(region) >= srcpos))
+    nb
```
```suggestion
    append!(dest.annotations, endannotations)
    nb
```

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+end
+
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}}) where {T <: AbstractString}
+    if position(io) == 0
+        AnnotatedString(read(io.io, T), copy(io.annotations))
+    else
+        start = position(io.io)
+        annots = [(max(1, first(region) - start):last(region)-start, val)
+                  for (region, val) in io.annotations if last(region) > start]
+        AnnotatedString(read(io.io, T), annots)
+    end
```
Can probably DRY this code a little bit:
```suggestion
    start = position(io.io)
    annotations = if start == 0
        copy(io.annotations)
    else
       typeof(io.annotations)[(max(1, first(region) - start):last(region)-start, val)
                  for (region, val) in io.annotations if last(region) > start]
    end
    return AnnotatedString(read(io.io, T), annotations)
```

tecosaur replied:
I'm not sure that this counts as DRY-er when you're only going from 44 to 42 symbols :sweat_smile:. If we want to go one half-step further, we can do `if (start = position(io)) == 0` (which I've just pushed).

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+pipe_reader(io::AnnotatedIOBuffer) = io.io
+pipe_writer(io::AnnotatedIOBuffer) = io.io
+
+# Useful `IOBuffer` methods that we don't get from `AbstractPipe`
```
Not to hold up this PR, but do you think we should add these to AbstractPipe also (all on the read side), except `copy` which doesn't have a generic implementation, or is it too much ambiguous whether this intends to attempt to manipulate the reader or writer half

tecosaur replied:
This is a slightly tricky one I think. I've found the lack of `position` and `seek` annoying with `IOContext`-wrapped `IOBuffer`s myself, however I think the problem is that sometimes IO has a read-head and a write-head, and sometimes it doesn't.

I think the cleaner way for this to be implemented is to have read-head and write-head variants of `position`/`seek`/etc. and if needed have them refer to the same head for IO types where there can only be one head. However, instead the default assumption is that they are the same, and so the model falls short as soon as there is distinct input and output IO.

vtjnash commented:
> I think that's everything addressed?

Yes, looked very good. I added a few more comments, mostly on the new code I think. Seems _very_ close to merging now.

vtjnash commented:
> With the way printstyled works, this benefit won't be seen immediately, but I see it as part of building the path to a nicer, less headache-inducing future

This should be straightforward though with this PR, no? So that existing users of printstyled can immediately see benefits for passing an `AnnotatedIOBuffer` instead, somewhat like the below code (warning: untested).

```
function printstyled(io::AnnotatedIOBuffer, things...; styles...)
    first = position(io)
    print(io, things...)
    last = position(io) - 1
    if last > first
        insert!(io.annotations, searchsortedfirst(io.annotations, first, by=last), (styles, first:last))
    end
    nothing
end
```
Or somewhat equivalantly:
```
function printstyled(io::AnnotatedIOBuffer, things...; styles...)
    write(io, AnnotatedString(sprint(things...; context=io), styles))
    nothing
end
```

vtjnash commented on base/strings/annotated.jl:
```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+            end
+            # Insert any extra entries in the appropriate position
+            for entry in extras
+                indices = searchsorted(io.annotations, (first(entry),), by=first)
+                splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
```
Do you mean to try to overwrite existing annotations here, or do you want to just insert the element at that place (which may not require as many temporary lists to be allocated):
```suggestion
                atwhere = searchsortedlast(io.annotations, (first(entry),), by=first)
                insert!(io.annotations, atwhere, entry)
```

tecosaur replied:
I indeed just want to insert annotations in the sorted location. Thanks!

tecosaur replied:
Ah, the `splice!` -> `insert!` change may not be viable after all (or needs a `max(1, _)`). Looking at the tests:

```
Error During Test at /home/tec/Documents/Projects/julia/test/strings/annotated.jl:3
  Got exception outside of a @test
  BoundsError: attempt to access 0-element Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}} at index [0]
  Stacktrace:
    [1] _growat!(a::Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}, i::Int64, delta::Int64)
      @ Base ./array.jl:1128
    [2] insert!
      @ ./array.jl:1727 [inlined]
    [3] annotate!(s::Base.AnnotatedString{String}, range::UnitRange{Int64}, labelval::Pair{Symbol})
      @ Base ./strings/annotated.jl:334
```

tecosaur replied:
Ah, seems like a `+ 1` is needed. Then the tests are happy, and so I've just pushed this version.

tecosaur commented:
>> With the way printstyled works, this benefit won't be seen immediately, but I see it as part of building the path to a nicer, less headache-inducing future
>
> This should be straightforward though with this PR, no? So that existing users of printstyled can immediately see benefits for passing an AnnotatedIOBuffer instead, somewhat like the below code (warning: untested).
>
> ```function printstyled(io::AnnotatedIOBuffer, things...; styles...)
>     first = position(io)
>     print(io, things...)
>     last = position(io) - 1
>     if last > first
>         insert!(io.annotations, searchsortedfirst(io.annotations, first, by=last), (styles, first:last))
>     end
>     nothing
> end
> ```
> 
> Or somewhat equivalantly:
> 
> ```
> function printstyled(io::AnnotatedIOBuffer, things...; styles...)
>     write(io, AnnotatedString(sprint(things...; context=io), styles))
>     nothing
> end
> ```

It is pretty straightforward indeed! That's why I've had https://github.com/JuliaLang/StyledStrings.jl/pull/12 sitting around since October, which lets you benefit from this new functionality with code that still uses `printstyled`. :smile:

Also, thanks for moving this forward through the review + merge process Jameson. It will be good to see this merged and start pushing on the PRs that this unlocks (including the currently open #51928 and #51914).

tecosaur committed:
Use insert! not splice! for new annotations

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

tecosaur commented:
Alrighty, time for round 2 of

> I think that's everything addressed?

:grin:

vtjnash commented:
Seems to have an ambiguity error:
```
Error in testset ambiguous:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-16a3361820/share/julia/test/ambiguous.jl:174
  Expression: sig1 === sig2
   Evaluated: read(s::IO, T::Type) @ Base io.jl:1179 === read(io::Base.AnnotatedIOBuffer, x) @ Base strings/annotated.jl:506
```

Also, I see you added the "don't squash" label, but it doesn't appear that each commit here is functionally independent. If not, they would normally be squashed to preserve the one commit == one feature / change expectation.

tecosaur commented:
>> It is pretty straightforward indeed! That's why I've had https://github.com/JuliaLang/StyledStrings.jl/pull/12 sitting around since October, which lets you benefit from this new functionality with code that still uses printstyled. 😄
> 
> Ah gotcha. That seems somewhat like type piracy though, no, since you have to load StyledStrings for this AnnotatedIOBuffer to work correctly?

It is indeed, type pirate-y. However, I don't really see much of a way around it.

One of the reasons (by a few people, IIRC) given in support of `StyledStrings` being a stdlib and not part of Base is not adding display/printing-logic-type code to Base when there are active efforts to slim it down (I belive somebody even made the aspirational comment that Base itself should have no printing code at all eventually). However, to work without causing a pile of method invalidations, some bits _need_ to be in Base (e.g. once this PR is merged, there's one that uses this to make `join` work better with a mix of `String`s and `AnnotatedString`s).

Now, `printstyled` is display/printing-logic, and so some of the functions required to convert it to a nice composable styled `AnnotatedString` are in the `StyledStrings` stdlib. So we have this split where some styling code is in `Base`, and some is in `StyledStrings`, and making the two work nicely together involves `StyledStrings` committing type piracy with `Base`  … but I don't see any easy way around it.  

Do please let me know if you have any better ideas around this.

vtjnash commented:
I guess that makes sense, maybe StyledStrings is just one of those that needs to be loaded by default, like Logging, FileWatching, and Libdl. They are conceptually separated out of Base, and could be removed during a build if absolutely required, but are not actually separable or "replaceable batteries" in that sense.

vtjnash commented:
Once the tests are passing, this looks good to merge

tecosaur commented:
I'm not actually quite sure what's going on with that ambiguity, might you be able to elaborate on it?

tecosaur commented:
> Also, I see you added the "don't squash" label, but it doesn't appear that each commit here is functionally independent. If not, they would normally be squashed to preserve the one commit == one feature / change expectation.

I guess it depends on what you look for in a PR. I get the sense that you're seeing this as a PR that "provides AnnotatedIOBuffer" and so there should be one commit that "introduces AnnotatedIOBuffer".

I tend to think of PRs as a collection of atomic commits that collectively provide a feature/change, and a useful history of how that happened. These commits are that.

```
* Introduce AnnotatedIOBuffer
  (An outline of the feature, with a commit message explaining the intent.)
* Make AnnotatedIOBuffer behave more like IOBuffer
  (Changing the behaviour of the feature, with justification in the commit message)
* Make AnnotatedIOBuffer reading more generic
  (Another specific change with justification/reasoning in the commit message)
* Also read AnnotatedChars from an AnnotatedIOBuffer
  (A new (sub)feature)
* Writing from one AnnotatedIOBuffer to another
  (A new (sub)feature)
* Tests for AnnotatedIOBuffer
* Update the mention of annotated features in NEWS
* Use insert! not splice! for new annotations
```

I'd think the history of the development (so long as the commits are atomic, which they are) would be more valuable than dumping the collection of changes fully formed into the repo, no? I guess it depends on how much you see git as a tool to track the development history of a project vs. just a means of distributed work.

Depending on how you look at it, forget being independent of each other, some of the commits here aren't even independent of other PRs :stuck_out_tongue: (e.g. the NEWS and insert! commits, affect content from previous PRs unrelated to this PR, but I wouldn't think are worth separating out into separate PRs).

tecosaur commented:
> I guess that makes sense, maybe StyledStrings is just one of those that needs to be loaded by default, like Logging, FileWatching, and Libdl

On this note, once #51869 is sorted out (imminently hopefully :crossed_fingers:), `StyledStrings` will be loaded with REPL, which I imagine is pretty similar in effect.

vtjnash commented:
> I'd think the history of the development

Commits should not reflect the history of the development, but rather are curated set of changes that each can be viewed independently as being a complete change

tecosaur commented:
> Commits should not reflect the history of the development, but rather are curated set of changes that each can be viewed independently as being a complete change

Right, but I don't see the two as being mutually exclusive — or do you not view commits like `Use insert! not splice! for new annotations` as being able to be viewed independently as a complete change?

vtjnash commented:
If it was in a separate PR, it would be a standalone commit, but as a complete sequence, each change to the API or implementation in the development process should not result in a separate commit to master

tecosaur commented:
You seem to be describing atomic commits, no?

The commits here are atomic, and can be handled individually. None of them are half-broken "aspirational" commits. The codebase is "valid" at each point, and CI should pass at each step.

tecosaur commented:
Feel free to just go ahead and squash merge this though, I care more about finally getting this PR over with than discussing whether it's worth preserving the commit history.

vtjnash commented:
A semi-automated way to look at it is that each separate commit should likely have tests if it represents a different change. Not always the case, but not all PRs need to have tests either. And if the later commits are changing the proposed API, then it is even less obvious that you would want someone to roll back this to a version with the older API as that would be a breaking change that wouldn't be allowed.

tecosaur commented:
I can't say I entirely see it, but this is good to merge from my POV. The CI hasn't completed yet, but the `buildkite/julia-master` only has a ``UndefVarError: `A` not defined in `Main.Test100Main_REPL.REPLCompletionsTest` `` failure so far. Hopefully the ambiguity check will clear this time.

tecosaur commented:
No ambiguity errors it looks like! :grinning:

tecosaur commented:
Thanks for helping this over the line Jameson!
